Modules/itertoolsmodule.c

```diff
@@ -51,6 +53,171 @@ static PyTypeObject pairwise_type;
 
 #include "clinic/itertoolsmodule.c.h"
 
+/* batched object ************************************************************/
+
+/* Note:  The built-in zip() function includes a "strict" argument
+   that is needed because that function can silently truncate data
+   and there is no easy way for a user to detect that condition.
+   The same reasoning does not apply to batches() which never drops
+   data.  Instead, it produces a shorter list which can be handled
+   as the user sees fit.
+ */
+
+typedef struct {
+    PyObject_HEAD
+    PyObject *it;
+    Py_ssize_t batch_size;
+} batchedobject;
+
+/*[clinic input]
+@classmethod
+itertools.batched.__new__ as batched_new
+    iterable: object
+    n: Py_ssize_t
+Batch data into lists of length n. The last batch may be shorter than n.
+
+Loops over the input iterable and accumulates data into lists
+up to size n.  The input is consumed lazily, just enough to
+fill a list.  The result is yielded as soon as a batch is full
+or when the input iterable is exhausted.
+
+    >>> for batch in batched('ABCDEFG', 3):
+    ...     print(batch)
+    ...
+    ['A', 'B', 'C']
+    ['D', 'E', 'F']
+    ['G']
+
+[clinic start generated code]*/
+
+static PyObject *
+batched_new_impl(PyTypeObject *type, PyObject *iterable, Py_ssize_t n)
+/*[clinic end generated code: output=7ebc954d655371b6 input=f28fd12cb52365f0]*/
+{
+    PyObject *it;
+    batchedobject *bo;
+
+    if (n < 1) {
+        /* We could define the n==0 case to return an empty iterator
+           but that is add odds with the idea that batching should
+           never throw-away input data.
+        */
+        PyErr_SetString(PyExc_ValueError, "n must be at least one");
+        return NULL;
+    }
+    it = PyObject_GetIter(iterable);
+    if (it == NULL) {
+        return NULL;
+    }
+
+    /* create batchedobject structure */
+    bo = (batchedobject *)type->tp_alloc(type, 0);
+    if (bo == NULL) {
+        Py_DECREF(it);
+        return NULL;
+    }
+    bo->batch_size = n;
+    bo->it = it;
+    return (PyObject *)bo;
+}
+
+static void
+batched_dealloc(batchedobject *bo)
+{
+    PyObject_GC_UnTrack(bo);
+    Py_XDECREF(bo->it);
+    Py_TYPE(bo)->tp_free(bo);
+}
+
+static int
+batched_traverse(batchedobject *bo, visitproc visit, void *arg)
+{
+    if (bo->it != NULL) {
+        Py_VISIT(bo->it);
+    }
+    return 0;
+}
+
+static PyObject *
+batched_next(batchedobject *bo)
+{
+    Py_ssize_t i;
+    PyObject *it = bo->it;
+    PyObject *item;
+    PyObject *result;
+
+    if (it == NULL) {
+        return NULL;
+    }
+    result = PyList_New(0);
```

Comment:
Maybe this could be `PyList_New(size)`? It would avoid most allocations, and `PyList_Append` could be replaced by `PyList_SET_ITEM` with less error handling. If wasting memory on the last batch is a concern, there could be a `Py_SET_SIZE(result, i);` followed by something like `Py_SETREF(result, PySequence_List(result));`.

I haven't benchmarked whether that makes a difference though.

Reply:
Another option might be `PyList_New(bo->batch_size);` followed by `Py_SET_SIZE(result, 0);`.

Reply:
My plan is to check in a straight-forward and obviously correct version.  Then you're welcome to make any optimizations to that baseline version.

ISTM that `PySequence_List()` is not guaranteed to work when the input contains `NULL` values.  The `PyList_New()` docs warn against going down that path.  It would be convenient if lists had some equivalent of `_PyTuple_Resize()`.